// protoscript/runtime/well-known-types/duration.pb.js
// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// Source: google/protobuf/duration.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "protoscript";
//========================================//
//        Protobuf Encode / Decode        //
//========================================//
export const Duration = {
    /**
     * Serializes Duration to protobuf.
     */
    encode: function (msg) {
        return Duration._writeMessage(msg, new BinaryWriter()).getResultBuffer();
    },
    /**
     * Deserializes Duration from protobuf.
     */
    decode: function (bytes) {
        return Duration._readMessage(Duration.initialize(), new BinaryReader(bytes));
    },
    /**
     * Initializes Duration with all fields set to their default value.
     */
    initialize: function () {
        return {
            seconds: 0n,
            nanos: 0,
        };
    },
    /**
     * @private
     */
    _writeMessage: function (msg, writer) {
        if (msg.seconds) {
            writer.writeInt64String(1, msg.seconds.toString());
        }
        if (msg.nanos) {
            writer.writeInt32(2, msg.nanos);
        }
        return writer;
    },
    /**
     * @private
     */
    _readMessage: function (msg, reader) {
        while (reader.nextField()) {
            const field = reader.getFieldNumber();
            switch (field) {
                case 1: {
                    msg.seconds = BigInt(reader.readInt64String());
                    break;
                }
                case 2: {
                    msg.nanos = reader.readInt32();
                    break;
                }
                default: {
                    reader.skipField();
                    break;
                }
            }
        }
        return msg;
    },
};
//========================================//
//          JSON Encode / Decode          //
//========================================//
export const DurationJSON = {
    /**
     * Serializes Duration to JSON.
     */
    encode: function (msg) {
        return JSON.stringify(DurationJSON._writeMessage(msg));
    },
    /**
     * Deserializes Duration from JSON.
     */
    decode: function (json) {
        return DurationJSON._readMessage(DurationJSON.initialize(), JSON.parse(json));
    },
    /**
     * Initializes Duration with all fields set to their default value.
     */
    initialize: function () {
        return {
            seconds: 0n,
            nanos: 0,
        };
    },
    /**
     * @private
     */
    _writeMessage: function (msg) {
        const json = {};
        if (msg.seconds) {
            json["seconds"] = msg.seconds.toString();
        }
        if (msg.nanos) {
            json["nanos"] = msg.nanos;
        }
        return json;
    },
    /**
     * @private
     */
    _readMessage: function (msg, json) {
        const _seconds = json["seconds"];
        if (_seconds) {
            msg.seconds = BigInt(_seconds);
        }
        const _nanos = json["nanos"];
        if (_nanos) {
            msg.nanos = _nanos;
        }
        return msg;
    },
};
